### ISSUE TITLE
How can I use one GPU to simulate over 100 concurrent clients and total 1000 clients in the pool? 

### ISSUE THREAD

**Issue by wizard1203** [adap/flower#1383]
### What is your question?

The flower paper says that the Virtual Client Engine can 

> enables large-scale single-machine or multi- machine experiments by executing workloads in a resource-aware fashion.

> It creates a ClientProxy for each client, but defers instantiation of the actual client object (including local model and data) until the resources to execute the client-side task (training, evaluation) become available.

Then I guess that flower can automatically schedule the simulation of different clients based on the resources. That means we can increase the number of clients largely, even infinite number of clients.

I tried 10 clients on one GPU, using the example https://github.com/adap/flower/blob/main/examples/advanced_pytorch/run.sh . But it cannot successfully run, with a CUDA out-of-memory error (16 GB memory). Could you please give example codes to implement this?  And even more clients than 10?

**Comment by mofanv** [comment]
I had the same problem `CUDA OOM`, but was running another script `simulation_pytorch`.
Can you try monitor how the OOM happened with a command like,
```
watch -n0.1 nvidia-smi
```
 to see if there are many `Ray::IDLE` processes occupying your GPU memory?

**Comment by wizard1203** [comment]
> I had the same problem `CUDA OOM`, but was running another script `simulation_pytorch`. Can you try monitor how the OOM happened with a command like,
> 
> ```
> watch -n0.1 nvidia-smi
> ```
> 
> to see if there are many `Ray::IDLE` processes occupying your GPU memory?

@mofanv Hi, I tried the script `simulation_pytorch`, it can successfully simulate 100 clients out of 1000 clients via one GPU, without OOM error. I'm trying to use more GPUs and more machines, do you know how to implement this? Like 2 machines with 8 GPUs, and 1000 concurrent clients out of 10000 clients (pool size).

**Comment by wizard1203** [comment]
It seems that indicating the num_gpus for each client can accelerate simulating, making fully utility of multiple GPUs?https://github.com/adap/flower/blob/aa9b9467fbb7b25d321ae0208434adc71e28bff2/baselines/flwr_baselines/publications/adaptive_federated_optimization/main.py#L32

**Comment by wizard1203** [comment]
@mofanv It seems that when I define
```
  client_resources = {
      "num_cpus": args.num_client_cpus,
      "num_gpus": 8,
  } 
```
The GPU memory will continuously increase. And there are many ray::IDLE process. 
If when I define 
```
  client_resources = {
      "num_cpus": args.num_client_cpus,
  } 
```
There is only one process occupying the GPU memory.

**Comment by jafermarq** [comment]
yes, you need to specify how much GPU memory a client will need. For example, if your client workload needs 4GB of VRAM (that you can see on nvidia-smi easily) and your GPU has 12GB of VRAM, then you should set `num_gpus: 0.3333`.

**Comment by jafermarq** [comment]
From ray docs: https://docs.ray.io/en/master/ray-core/tasks/resources.html. So, if you set `client_resources = {num_cpus:4, num_gpus:0.5}` and pass it to `start_simulation()`, then each client will use no more than `4` of your CPU threads and ray **will assume** the client uses at most 50% of your GPU. This is an assumption so it's up to the developer to set that `num_gpus` correctly

**Comment by wizard1203** [comment]
> yes, in the same way you can set `num_cpus`, you need to specify how much GPU memory a client will need. For example, if your client workload needs 4GB of VRAM (that you can see on nvidia-smi easily) and your GPU has 12GB of VRAM, then you should set `num_gpus: 0.3333`.

@jafermarq Thanks for your kind reply. I hope to know how to efficiently utilize all 8 GPUs to simulate over than 100 even 1000 clients (Using the `fl.client.start_numpy_client` to launch 1000 clients is not practical because the memory constraint.). It seems that the ray cannot implement such an efficient parallel training on multiple GPUs?

**Comment by jafermarq** [comment]
Hi @wizard1203 , Flower can train efficiently with 1000+ clients. My advice is that you try Flower's simulation mode instead (which uses Ray under the hood to manage the resources in your system and schedule clients based on: (1) your resources and (2) the resources each client needs). This is the PyTorch example: https://github.com/adap/flower/tree/main/examples/simulation_pytorch

That example runs clients on CPU, but if you want them to use GPU, you need to edit this line: https://github.com/adap/flower/blob/aa9b9467fbb7b25d321ae0208434adc71e28bff2/examples/simulation_pytorch/main.py#L142

and add `num_gpus: X`, where `X` is fractional number (e.g. 0.5). This means that a client will use an x-th of a GPU. if x=0.5, each client will be allocated 50% of GPU memory in  your system. if you set x=2, each client will get 2 GPUs

**Comment by jafermarq** [comment]
I can put together an example later in the day and share it here

**Comment by wizard1203** [comment]
> Hi @wizard1203 , Flower can train efficiently with 1000+ clients. My advice is that you try Flower's simulation mode instead (which uses Ray under the hood to manage the resources in your system and schedule clients based on: (1) your resources and (2) the resources each client needs). This is the PyTorch example: https://github.com/adap/flower/tree/main/examples/simulation_pytorch
> 
> That example runs clients on CPU, but if you want them to use GPU, you need to edit this line:
> 
> https://github.com/adap/flower/blob/aa9b9467fbb7b25d321ae0208434adc71e28bff2/examples/simulation_pytorch/main.py#L142
> 
> and add `num_gpus: X`, where `X` is fractional number (e.g. 0.5). This means that a client will use an x-th of a GPU. if x=0.5, each client will be allocated 50% of GPU memory in your system. if you set x=2, each client will get 2 GPUs

Yeah, I'm trying this example. But I'm confused about this line https://github.com/adap/flower/blob/aa9b9467fbb7b25d321ae0208434adc71e28bff2/examples/simulation_pytorch/main.py#L32, which indicates the client only uses the cuda:0 device. I can see all GPUs are used. Does it means that the ray will automatically schedule all devices to these clients? The cuda:0 for clients are actually cuda0\~7. Clients can only see one device, which is assigned by the ray.

**Comment by wizard1203** [comment]
> I can put together an example later in the day and share it here

I'm really appreciated that if you could provide one more example which uses multiple machines.

**Comment by jafermarq** [comment]
The way to think about `self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu") ` is as follows: Ray masks out all but the GPU assigned to a client, so a client can only "see" the GPU it got assigned. So that would be first GPU (i.e. `cuda:0`)

**Comment by jafermarq** [comment]
> > I can put together an example later in the day and share it here
> 
> I'm really appreciated that if you could provide one more example which uses multiple machines.

This can be done already by passing the correct arguments to `ray_init_args` in `fl.simulation.start_simulation`. I'll share an example later.

**Comment by jafermarq** [comment]
So these are the steps to run a multi-node Flower simulation. 

0. Ensure you have the exact same code base, environment (exact same python version too), and data partitions in each node.
1. Open a terminal on one of the nodes (this will be your head node orchestrating the whole distributed FL experiment) and type in the terminal: `ray start --head`. This should print something as follows:
```
Local node IP: 192.168.1.84

--------------------
Ray runtime started.
--------------------

Next steps
  To connect to this Ray runtime from another node, run
    ray start --address='192.168.1.84:6379' --redis-password='5241590000000000'
  
  Alternatively, use the following Python code:
    import ray
    ray.init(address='auto', _redis_password='5241590000000000')
  
  To connect to this Ray runtime from outside of the cluster, for example to
  connect to a remote cluster from your laptop directly, use the following
  Python code:
    import ray
    ray.init(address='ray://<head_node_ip_address>:10001')
  
  If connection fails, check your firewall settings and network configuration.
  
  To terminate the Ray runtime, run
    ray stop
```
2. From the message printed you can see what you need to do in each of the other nodes. In this case: open a terminal and type `ray start --address='192.168.1.84:6379' --redis-password='5241590000000000'` in order for this node to connect with the head node. Then repeat this for the remaining nodes.
3. Go back to your head now and you can type in the terminal `ray status` and information about each node connected to the head. In my case I used two nodes so I see:
```
======== Autoscaler status: 2022-08-24 21:13:15.548859 ========                                                                                                                         
Node status                                                                                                                                                                             
---------------------------------------------------------------                                                                                                                         
Healthy:                                                                                                                                                                                
 1 node_01d9f309d67e833ce73f896799e40e5eadea605cf37ebaf1fb625c66                                                                                                                        
 1 node_43df4ec806d8f7824c2c76c12bee564fc691a4436904229789091c34                                                                                                                        
Pending:                                                                                                                                                                                
 (no pending nodes)                                                                                                                                                                     
Recent failures:                                                                                                                                                                        
 (no failures)                                                     
 ```
4. In your head node and in your python code ensure ray is initialised with `address="auto"`. This will tell ray to search for an already running server in this (the head) node. In Flower (e.g. in the [`simulation_pytorch`]() example) you need to ensure that the `ray_init_args` argument to `fl.simulation.start_simulation()` specifies  `address="auto"`.
5. All is set! from your head node run the experiment. The head node will spawn clients in in other nodes (if not all fit at the same time in the head node). Ray internally does all the communication between remote nodes and your head nodes so all is handled for you.

Let me know if you get it working !
